File: backend/src/services/files/pdf.ts
import { PDFParse } from "pdf-parse";

/**
 * Extracts plain text from a PDF buffer.
 *
 * Uses pdf-parse v2's class-based API (`PDFParse`/`getText`), which is a
 * ground-up rewrite of the legacy 1.x `pdf-parse(buffer)` function API.
 * Never throws: any failure (malformed PDF, image-only PDF, parser error)
 * results in an empty string so callers can safely fall back.
 */
export async function parsePdf(buffer: Buffer): Promise<string> {
  let parser: PDFParse | undefined;
  try {
    parser = new PDFParse({ data: buffer });
    const result = await parser.getText();
    return result.text.trim().slice(0, 60_000);
  } catch {
    return "";
  } finally {
    if (parser) {
      try {
        await parser.destroy();
      } catch {
        // ignore cleanup errors
      }
    }
  }
}

File: backend-py/app/services/files/pdf.py
from __future__ import annotations

import io
import logging

from pypdf import PdfReader

from app.services.files._common import parse_in_thread

logger = logging.getLogger(__name__)


def _extract_text(buffer: bytes) -> str:
    reader = PdfReader(io.BytesIO(buffer))
    text = "\n".join(page.extract_text() or "" for page in reader.pages)
    return text.strip()[:60_000]


async def parse_pdf(buffer: bytes) -> str:
    """Extracts plain text from a PDF buffer. Never throws: any failure
    (malformed PDF, image-only PDF, parser error) results in an empty
    string so callers can safely fall back.

    Parsing is synchronous/CPU-bound (pypdf has no async API), so it runs
    on a worker thread via `asyncio.to_thread` to avoid blocking the event
    loop for the duration of a large upload's parse.
    """
    return await parse_in_thread(_extract_text, buffer, logger, "parse_pdf")
